File: app/src/main/java/jp/juggler/subwaytooter/util/EmojiImageSpan.java
package jp.juggler.subwaytooter.util;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.drawable.Drawable;
import android.support.annotation.IntRange;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
//import android.text.style.DynamicDrawableSpan;
import android.text.style.ReplacementSpan;

import java.lang.ref.WeakReference;

public class EmojiImageSpan extends ReplacementSpan {
	
	// private static final LogCategory log = new LogCategory( "EmojiImageSpan" );

	// static DynamicDrawableSpan x = null;
	
	private static final float scale_ratio = 1.14f;
	private static final float descent_ratio = 0.211f;
	
	private final Context context;
	private final int res_id;
	private WeakReference< Drawable > mDrawableRef;
	
	public EmojiImageSpan( @NonNull Context context, int res_id ){
		super();
		this.context = context.getApplicationContext();
		this.res_id = res_id;
	}
	
	@Override
	public int getSize(
		@NonNull Paint paint
		, CharSequence text
		, @IntRange(from = 0) int start
		, @IntRange(from = 0) int end
		, @Nullable Paint.FontMetricsInt fm
	){
		int size = (int) ( 0.5f + scale_ratio * paint.getTextSize() );
		
		if( fm != null ){
			int c_descent = (int) ( 0.5f + size * descent_ratio );
			int c_ascent = c_descent - size;
			if( fm.ascent > c_ascent ) fm.ascent = c_ascent;
			if( fm.top > c_ascent ) fm.top = c_ascent;
			if( fm.descent < c_descent ) fm.descent = c_descent;
			if( fm.bottom < c_descent ) fm.bottom = c_descent;
		}
		return size;
	}
	
	@Override public void draw(
		@NonNull Canvas canvas
		, CharSequence text, int start, int end
		, float x, int top, int baseline, int bottom
		, @NonNull Paint paint
	){
		int size = (int) ( 0.5f + scale_ratio * paint.getTextSize() );
		int c_descent = (int) ( 0.5f + size * descent_ratio );
		
		Drawable d = getCachedDrawable();
		d.setBounds( 0, 0, size, size );
		
		int transY = baseline - size + c_descent;
		
		canvas.save();
		canvas.translate( x, transY );
		d.draw( canvas );
		canvas.restore();
	}
	
	private Drawable getCachedDrawable(){
		Drawable d = null;
		
		if( mDrawableRef != null ){
			d = mDrawableRef.get();
		}
		
		if( d == null ){
			d = ContextCompat.getDrawable( context, res_id );
			mDrawableRef = new WeakReference<>( d );
		}
		
		return d;
	}
	
}

File: app/src/main/java/jp/juggler/subwaytooter/span/EmojiImageSpan.kt
package jp.juggler.subwaytooter.span

import android.content.Context
import android.graphics.*
import android.graphics.drawable.Drawable
import android.text.style.ReplacementSpan
import androidx.annotation.IntRange
import androidx.core.content.ContextCompat
import java.lang.ref.WeakReference

class EmojiImageSpan(
    context: Context,
    private val resId: Int,
    private val useColorShader: Boolean = false,
    private val color: Int? = null,
    private val scale: Float = 1f
) : ReplacementSpan() {

    companion object {
        // private static final LogCategory warning = new LogCategory( "EmojiImageSpan" );
        // static DynamicDrawableSpan x = null;
        private const val scaleRatio = 1.14f
        private const val descentRatio = 0.211f
    }

    private val context: Context
    private var mDrawableRef: WeakReference<Drawable>? = null

    private val cachedDrawable: Drawable?
        get() {
            var d = mDrawableRef?.get()
            if (d == null) {
                d = ContextCompat.getDrawable(context, resId) ?: return null
                mDrawableRef = WeakReference(d)
            }
            return d
        }

    init {
        this.context = context.applicationContext
    }

    private fun getImageSize(paint: Paint) = (0.5f + scaleRatio * scale * paint.textSize).toInt()

    override fun getSize(
        paint: Paint,
        text: CharSequence,
        @IntRange(from = 0) start: Int,
        @IntRange(from = 0) end: Int,
        fm: Paint.FontMetricsInt?
    ): Int {

        val size = getImageSize(paint)

        if (fm != null) {
            val cDescent = (0.5f + size * descentRatio).toInt()
            val cAscent = cDescent - size
            if (fm.ascent > cAscent) fm.ascent = cAscent
            if (fm.top > cAscent) fm.top = cAscent
            if (fm.descent < cDescent) fm.descent = cDescent
            if (fm.bottom < cDescent) fm.bottom = cDescent
        }
        return size
    }

    private var lastColor: Int? = null
    private var lastColorFilter: PorterDuffColorFilter? = null

    override fun draw(
        canvas: Canvas,
        text: CharSequence,
        start: Int,
        end: Int,
        x: Float,
        top: Int,
        baseline: Int,
        bottom: Int,
        paint: Paint
    ) {
        val d = cachedDrawable ?: return

        val size = getImageSize(paint)
        val cDescent = (0.5f + size * descentRatio).toInt()
        val transY = baseline - size + cDescent

        canvas.save()
        canvas.translate(x, transY.toFloat())
        d.setBounds(0, 0, size, size)

        if (useColorShader) {
            val pc: Int
            val pa: Int
            if (color == null) {
                pc = paint.color
                pa = paint.alpha
            } else {
                pc = color or Color.BLACK
                pa = color ushr 24
            }
            // Log.d("EmojiImageSpan",String.format("paint c=0x%x a=0x%x",pc,pa))
            if (pc != lastColor || lastColorFilter == null) {
                lastColor = pc
                lastColorFilter = PorterDuffColorFilter(pc or Color.BLACK, PorterDuff.Mode.SRC_ATOP)
            }
            val saveColorFilter = d.colorFilter
            val saveAlpha = d.alpha
            d.colorFilter = lastColorFilter
            d.alpha = pa
            d.draw(canvas)
            d.colorFilter = saveColorFilter
            d.alpha = saveAlpha
        } else {
            d.draw(canvas)
        }
        canvas.restore()
    }
}
